README.md
Lingua-TFIDF
===

This is just another TF-IDF calculator for Perl 5.

Quoting [Wikipedia](http://en.wikipedia.org/wiki/Tf%E2%80%93idf):

> tf–idf, short for term frequency–inverse document frequency, is a numerical statistic that is intended to reflect how important a word is to a document in a collection or corpus. It is often used as a weighting factor in information retrieval and text mining.

I know that CPAN has several TF-IDF calculator modules already, like Text::TFIDF and Lingua::JA::TFIDF. So why I reinvented the wheel? The reason is language dependency: Text::TFIDF assumes that words in sentence are separated by spaces in spite of that the assumption is not true in most east asian languages. And Lingua::JA::TFIDF works only on documents written in Japanese.

Lingua::TFIDF solves this problem by separating the process into 2 parts: word segmentation and word frequency counting. By supplying a language-specific word segmenter, this class can calculate TF-IDF on documents written in any languages.

Example
---

```perl
#!/usr/bin/env perl

use strict;
use warnings;
use Lingua::TFIDF;
use Lingua::TFIDF::WordSegmenter::SplitBySpace;

my $tf_idf_calc = Lingua::TFIDF->new(
  word_segmenter => Lingua::TFIDF::WordSegmenter::SplitBySpace->new,
);

my @documents = (<<'EOD1', <<'EOD2');
Humpty Dumpty sat on a wall,
Humpty Dampty had a great fall.
...
EOD1
Remember, remember, the fifth of November,
The Gunpowder Treason and plot.
...
EOD2

my $tf = $tf_idf_calc->tf(document => $documents[0], normalize => 1);
my $idf = $tf_idf_calc->idf(documents => \@documents);
my $tf_idf = $tf_idf_calc->tf_idf(documents => \@documents, normalize => 1);
```

License
---

The MIT License (MIT)

Copyright (c) 2014 Koichi SATOH, all rights reserved.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


Changes
Revision history for Lingua-TFIDF

{{$NEXT}}

0.01    2014-05-12
        - Initial release.


lib/Lingua/TFIDF/WordSegmenter/JA/MeCab.pm
package Lingua::TFIDF::WordSegmenter::JA::MeCab;

# ABSTRACT: Word segmenter for Japanese documents

use strict;
use warnings;
use Encode qw//;
use Smart::Args;
use Text::MeCab;

my $mecab_encoding = Encode::find_encoding(Text::MeCab::ENCODING);

sub new {
  args
    my $class => 'ClassName',
    my $mecab => +{ isa => 'Text::MeCab', optional => 1 };

  $mecab = Text::MeCab->new unless defined $mecab;
  bless +{ mecab => $mecab } => $class;
}

sub mecab { $_[0]->{mecab} }

sub segment {
  args_pos
    my $self,
    my $document => 'Ref | Str';

  my $input = $mecab_encoding->encode(ref $document ? $$document : $document);
  my $node = $self->mecab->parse($input);
  sub {
    return unless $node and $node->stat != Text::MeCab::MECAB_EOS_NODE;
    my $word = $mecab_encoding->decode($node->surface);
    $node = $node->next;
    return $word;
  };
}

1;

=head1 SYNOPSIS

  use utf8;
  use Lingua::TFIDF::WordSegmenter::JA::MeCab;
  
  my $segmenter = Lingua::TFIDF::WordSegmenter::JA::MeCab->new;
  my $iter = $segmenter->segment('思い出せ、思い出せ　11月5日を...');
  while (defined(my $word = $iter->())) { ... }

=head1 DESCRIPTION

This class is a word segmenter for documents written in Japanese.

=head1 METHODS

=head2 new([ mecab => Text::MeCab->new ])

Constructor.

=head2 segment($document | \$document)

Executes word segmentation on given C<$document> and returns an word iterator.

=head1 SEE ALSO

L<Text::MeCab>

=cut


lib/Lingua/TFIDF/WordSegmenter/LetterNgram.pm
package Lingua::TFIDF::WordSegmenter::LetterNgram;

# ABSTRACT: Letter N-gram segmenter

use strict;
use warnings;
use Carp qw//;
use Smart::Args;

sub new {
  args
    my $class => 'ClassName',
    my $n => 'Int';

  Carp::croak('Word length must be 1+.') if $n <= 0;

  bless +{ n => $n } => $class;
}

sub n { $_[0]->{n} }

sub segment {
  args_pos
    my $self,
    my $document => 'Ref | Str';

  $document = \"$document" unless ref $document;

  my $length = length $$document;
  my $index = -1;
  my $n = $self->n;
  sub {
  GET_NEXT_NGRAM:
    {
      ++$index;
      return if $index + $n > $length;

      my $ngram = substr $$document, $index, $n;
      redo GET_NEXT_NGRAM if $ngram =~ /\s/;
      return $ngram;
    }
  };
}

1;

=head1 SYNOPSIS

  use Lingua::TFIDF::WordSegmenter::LetterNgram;
  
  my $segmenter = Lingua::TFIDF::WordSegmenter::LetterNgram->new(n => 2);
  my $iter = $segmenter->segment('ロンドン橋落ちた　落ちた　落ちた...');
  while (defined(my $word = $iter->())) { ... }

=head1 DESCRIPTION

This class provides a N-gram word segmenter.

=head1 METHODS

=head2 new(n => $n)

Constructor.

=head2 segment($document | \$document)

Executes word segmentation on given C<$document> and returns an word iterator.

=cut


lib/Lingua/TFIDF/WordSegmenter/SplitBySpace.pm
package Lingua::TFIDF::WordSegmenter::SplitBySpace;

# ABSTRACT: Simple word segmenter suitable for most european languages

use strict;
use warnings;
use Smart::Args;

sub new {
  args
    my $class => 'ClassName',
    my $lower_case => +{ isa => 'Bool', default => 0 },
    my $remove_punctuations => +{ isa => 'Bool', default => 0 },
    my $stop_words => +{ isa => 'ArrayRef[Str]', default => [] };

  bless +{
    lower_case => $lower_case,
    remove_punctuations => $remove_punctuations,
    stop_words => $stop_words,
  } => $class;
}

sub lower_case { $_[0]->{lower_case} }

sub remove_punctuations { $_[0]->{remove_punctuations} }

sub segment {
  args_pos
    my $self,
    my $document => 'Ref | Str';

  my @words = split /\s+/, ref $document ? $$document : $document;

  @words = map lc, @words if $self->lower_case;

  if ($self->remove_punctuations) {
    s/^\W+|\W+$//g for @words;
  }

  if (@{ $self->stop_words } != 0) {
    my %stop_words = map { ($_ => 1) } @{ $self->stop_words };
    @words = grep { not exists $stop_words{$_} } @words;
  }

  return sub { shift @words };
}

sub stop_words { $_[0]->{stop_words} }

1;

=head1 SYNOPSIS

  use Lingua::TFIDF::WordSegmenter::SplitBySpace;
  
  my $segmenter = Lingua::TFIDF::WordSegmenter::SplitBySpace->new(
    lower_case => 1,
    remove_punctuations => 1,
    stop_words => [qw/i you he she it they a the am are is was were/],
  );
  my $iter = $segmenter->segment('Humpty Dumpty sat on wall, ...');
  while (defined(my $word = $iter->())) { ... }

=head1 DESCRIPTION

This class is a simple word segmenter. Like L<Text::TFIDF>, this class segments a sentence into words by spliting by spaces.

=head1 METHODS

=head2 new([ lower_case => 0 ] [, remove_punctuations => 0 ] [, stop_words => [] ])

Constructor. Takes some optional parameters:

=over 2

=item lower_case

Set off by default. Convert all the words into lower cases.

=item remove_punctuations

Set off by default. Removes punctuation characters (e.g., commas, periods, quotes, question marks and exclamation marks) from head and tail of segmented words. Note that punctuations at inside of a word (e.g., "King's") will be remain unchanged.

=item stop_words

Specifies words you want to exclude from segmented words. This is useful for removing functional words.

Note that stop word filtering will be performed B<after> C<lower_case> and C<remove_punctuations> options are processed. So, for example, if you enable C<lower_case> option and want to exclude "I" from result, you should supply the stop word list as C<['i']>.

=back

=head2 segment($document | \$document)

Executes word segmentation on given C<$document> and returns an word iterator.

=cut


lib/Lingua/TFIDF.pm
package Lingua::TFIDF;

# ABSTRACT: Language-independent TF-IDF calculator.

use strict;
use warnings;
use Lingua::TFIDF::Types;
use List::MoreUtils qw/uniq/;
use List::Util qw/sum/;
use Smart::Args;

our $VERSION = 0.01;

sub new {
  args
    my $class => 'ClassName',
    my $word_counter => +{ isa => 'Lingua::TFIDF::WordCounter', optional => 1 },
    my $word_segmenter => 'Lingua::TFIDF::WordSegmenter';

  unless (defined $word_counter) {
    require Lingua::TFIDF::WordCounter::Simple;
    $word_counter = Lingua::TFIDF::WordCounter::Simple->new;
  }

  bless +{
    word_counter => $word_counter,
    word_segmenter => $word_segmenter,
  } => $class;
}

sub idf {
  args
    my $self,
    my $documents => 'ArrayRef[Lingua::TFIDF::TermFrequency] | ArrayRef[Str]';

  return +{} if @$documents == 0;

  my @tfs = ref $documents->[0]
    ? @$documents : map { $self->tf(document => \$_) } @$documents;
  my %idf;
  for my $word (uniq map { keys %$_ } @tfs) {
    my $num_documents_including_word = grep { exists $_->{$word} } @tfs;
    $idf{$word} = log(@tfs / $num_documents_including_word);
  }
  return \%idf;
}

sub tf {
  args
    my $self,
    my $document => 'Ref | Str',
    my $normalize => +{ isa => 'Bool', default => 0 };

  $self->word_counter->clear;

  my $iter = $self->word_segmenter->segment($document);
  my $counter = $self->word_counter;
  while (defined (my $word = $iter->())) { $counter->add_count($word) }

  my $tf = $counter->frequencies;
  return $tf unless $normalize;

  my $total_words = sum values %$tf;
  +{ map { ($_ => $tf->{$_} / $total_words) } keys %$tf };
}

sub tf_idf {
  args
    my $self,
    my $documents => 'ArrayRef[Str]',
    my $normalize => +{ isa => 'Bool', default => 0 };

  return +{} if @$documents == 0;

  my @tfs =
    map { $self->tf(document => \$_, normalize => $normalize) } @$documents;
  my $idf = $self->idf(documents => \@tfs);
  my @tf_idf;
  for my $tf (@tfs) {
    push @tf_idf, +{ map { ($_ => $tf->{$_} * $idf->{$_}) } keys %$tf };
  }
  return \@tf_idf;
}

sub word_counter { $_[0]->{word_counter} }

sub word_segmenter { $_[0]->{word_segmenter} }

1;

=head1 SYNOPSIS

  use Lingua::TFIDF;
  use Lingua::TFIDF::WordSegmenter::SplitBySpace;
  
  my $tf_idf_calc = Lingua::TFIDF->new(
    word_segmenter => Lingua::TFIDF::WordSegmenter::SplitBySpace->new,
  );
  
  my $document1 = 'Humpty Dumpty sat on a wall...';
  my $document2 = 'Remember, remember, the fifth of November...';
  
  my $tf = $tf_idf_calc->tf(document => $document1);
  # TF of word "Dumpty" in $document1.
  say $tf->{'Dumpty'};  # 2, if you are referring same text as mine.
  
  my $idf = $tf_idf_calc->idf(documents => [$document1, $document2]);
  say $idf->{'Dumpty'};  # log(2/1) ≒ 0.693147
  
  my $tf_idfs = $tf_idf_calc->tf_idf(documents => [$document1, $document2]);
  # TF-IDF of word "Dumpty" in $document1.
  say $tf_idfs->[0]{'Dumpty'};  # 2 log(2/1) ≒ 1.386294
  # Ditto. But in $document2.
  say $tf_idfs->[1]{'Dumpty'};  # 0

=head1 DESCRIPTION

Quoting L<Wikipedia|http://en.wikipedia.org/wiki/Tf%E2%80%93idf>:

  tf–idf, short for term frequency–inverse document frequency, is a numerical statistic that is intended to reflect how important a word is to a document in a collection or corpus. It is often used as a weighting factor in information retrieval and text mining.

This module provides feature for calculating TF, IDF and TF-IDF.

=head2 MOTIVATION

There are several TF-IDF calculator modules in CPAN already, for example L<Text::TFIDF> and L<Lingua::JA::TFIDF>. So why I reinvent the wheel? The reason is language dependency: C<Text::TFIDF> assumes that words in sentence are separated by spaces. This assumption is not true in most east asian languages. And C<Lingua::JA::TFIDF> works only on japanese text.

C<Lingua::TFIDF> solves this problem by separating word segmentation process from word frequency counting. You can process documents written in any languages, by providing appropriate word segmenter (see L</CUSTOM WORD SEGMENTER> below.)

=head1 METHODS

=head2 new(word_segmenter => $segmenter)

Constructor. Takes 1 mandatory parameter C<word_segmenter>.

=head3 CUSTOM WORD SEGMENTER

Although this distribution bundles some language-independent word segmenter, like L<Lingua::TFIDF::WordSegmenter::SplitBySpace>, sometimes language-specifiec word segmenters are more appropriate. You can pass a custom word segmenter object to the calculator.

The word segmenter is a plain Perl object that implements C<segment> method. The method takes 1 positional argument C<$document>, which is a string or a B<reference> to string. It is expected to return an word iterator as CodeRef.

Roughly speaking, given custom word segmenter will be used like:

  my $document = 'foo bar baz';
  
  # Can be called with a reference, like |->segment(\$document)|.
  # Detecting data type is callee's responsibility.
  my $iter = $word_segmenter->segment($document);
  
  while (defined(my $word = $iter->())) {
     ...
  }

=head2 idf(documents => \@documents)

Calculates IDFs. Result is returned as HashRef, which the keys and values are words and corresponding IDFs respectively.

=head2 tf(document => $document | \$document [, normalize => 0])

Calculates TFs. Result is returned as HashRef, which the keys and values are words and corresponding TFs respectively.

If optional parameter <normalize> is set true, the TFs are devided by the number of words in the C<$document>. It is useful when comparing TFs with other documents.

=head2 tf_idf(documents => \@documents [, normalize => 0])

Calculates TF-IDFs. Result is returned as ArrayRef of HashRef. Each HashRef contains TF-IDF values for corresponding document.

=head1 SEE ALSO

=over 2

=item L<Lingua::TFIDF::WordSegmenter::LetterNgram>

=item L<Lingua::TFIDF::WordSegmenter::SplitBySpace>

=item L<Lingua::TFIDF::WordSegmenter::JA::MeCab>

=back

=cut
